src/workers/newNotificationV2Push.ts

```diff
@@ -0,0 +1,60 @@
+import { messageToJson, Worker } from './worker';
+import { ChangeObject } from '../types';
+import { NotificationV2 } from '../entity';
+import { sendPushNotification } from '../onesignal';
+import { getDisconnectedUsers } from '../subscription';
+import { processStreamInBatches } from '../common/streaming';
+import {
+  getNotificationV2AndChildren,
+  streamNotificationUsers,
+} from '../notifications/common';
+
+interface Data {
+  notification: ChangeObject<NotificationV2>;
+}
+
+const BATCH_SIZE = 100;
+const QUEUE_CONCURRENCY = 10;
+
+const worker: Worker = {
+  subscription: 'api.new-notification-push',
+  handler: async (message, con, logger): Promise<void> => {
+    const data: Data = messageToJson(message);
+    if (data.notification.public) {
+      const [notification, , avatars] = await getNotificationV2AndChildren(
+        con,
+        data.notification.id,
+      );
+      if (notification) {
+        try {
+          const stream = await streamNotificationUsers(con, notification.id);
+          await processStreamInBatches(
+            stream,
+            async (batch: { userId: string }[]) => {
+              const users = await getDisconnectedUsers(
+                batch.map((b) => b.userId),
+              );
+              if (users.length) {
+                await sendPushNotification(users, notification, avatars?.[0]);
+              }
+            },
+            QUEUE_CONCURRENCY,
+            BATCH_SIZE,
+          );
+        } catch (err) {
+          // Don't throw the error because we send the notification to multiple users, retrying may spam some users
+          logger.error(
```

Comment:
Would splitting this in more smaller jobs fix the issue with spam eg. this main one can publish a message to send each one. It would increase number of messages but if we moved it to separate deployment it would scale separately?

Reply:
https://github.com/dailydotdev/daily-api/pull/1582#discussion_r1423899170
To elaborate, on a nutshell you're right. But the tradeoff is consistency vs latency. So introducing another layer of complexity will increase the lag between the event and push and ideally I'd like to minimize that as much as possible

Reply:
Yeah got it, the above would also help with personalizing notifications. I am fine with leaving it for now, don't think it would be much work to split if needed down the line 👍